bundles/org.eclipse.equinox.p2.core/src/org/eclipse/equinox/internal/p2/core/ProvisioningAgent.java

```diff
@@ -26,13 +26,13 @@
 /**
  * Represents a p2 agent instance.
  */
-public class ProvisioningAgent implements IProvisioningAgent, ServiceTrackerCustomizer<IAgentServiceFactory, Object> {
+public class ProvisioningAgent implements IProvisioningAgent {
 
 	private final Map<String, Object> agentServices = Collections.synchronizedMap(new HashMap<>());
 	private BundleContext context;
 	private volatile boolean stopped = false;
 	private ServiceRegistration<IProvisioningAgent> reg;
-	private final Map<ServiceReference<IAgentServiceFactory>, ServiceTracker<IAgentServiceFactory, Object>> trackers = Collections
+	private final Map<String, ServiceTracker<IAgentServiceFactory, Object>> trackers = Collections
```

Comment:
I would here today use a `ConcurrentHashMap`, also instead of storing a `ServiceTracker`  object it would be better to use a dedicated class (that internal holds / manages a `ServiceTracker`), then one can use a quite nice pattern in a way that one first computes that class and then sync on the methods of that particular class. That way the map can work completely lock-free.